macros/get_metric_sql.sql

```diff
@@ -4,19 +4,55 @@
       - allow start/end dates on metrics. Maybe special-case "today"?
       - allow passing in a seed with targets for a metric's value
 */
-{%- macro get_metric_sql(metric_list, grain, dimensions, secondary_calculations, start_date, end_date, where) %}
+{%- macro get_metric_sql(metric_list, grain, dimensions, secondary_calculations, start_date, end_date, where, initiated_by,metric_definition=None) %}
 
+{# ############
+VALIDATION AROUND METRIC VS CALCULATE VS DEVELOP
+############ #}
+
+{% if initiated_by == 'calculate' %}
+    {% set is_calculate_macro = true %}
+    {% set is_develop_macro = false %}
+{% elif initiated_by == 'develop' %}
+    {% set is_calculate_macro = false %}
+    {% set is_develop_macro = true %}
+{% endif %}
+
+{# ############
+METRIC ATTRIBUTES TO VARIABLES
+We do this so that we can support metric attributes coming in from other formats
+like develop. This is easier than altering all of the downstream logic to have flags
+for calculate vs develop
+############ #}
+{% if is_develop_macro %}
+    {% set metric_name = metric_definition["name"]%}
+    {% set metric_type = metric_definition["type"]%}
+    {% set metric_sql = metric_definition["sql"]%}
+    {% set metric_timestamp = metric_definition["timestamp"]%}
+    {% set metric_time_grains = metric_definition["time_grains"]%}
+    {% set metric_dimensions = metric_definition["dimensions"]%}
+    {% set metric_filters = metric_definition["filters"]%}
+    {% set metric_base_model = metric_definition["model"].replace('"','\'').split('\'')[1]  %}
+    {% set metric_model = metrics.get_model_relation(metric_base_model if execute else "") %}
+{% endif %}
```

Comment:
Could you create a valid metric object instead? This feels very clunky. In particular, if at all possible I'd like it if the get_metric_sql could remain blissfully unaware of the existence of the develop macro. 

This leads to very tightly coupled functions as opposed to the get_metric_sql macro exposing a fixed interface that things that want to integrate with it have to conform to.

Reply:
Per my conversations with @jtcohen6 , I believe it is impossible to temporarily create a metric object.

> We don't have any mechanism to create a real metrics node without updating project code + triggering a re-parse. (Something to think about!!!)
So - I think the right move is just to hack the inputs to the metrics query compilation, by pulling the needed fields out of the user-supplied fromyaml-derived dictionary

In the future we might be able to abstract this more but right now I think that is not technically feasible

Reply:
Yeah, no way to programmatically create a new object in the manifest at runtime... maybe in dbt v2? :)

Reply:
OK, phrased a different way: does a metric node contain any methods/properties that you need to use inside of get_metric_sql? I'm not entirely sure whether it contains _any_ (thinking about like how you can do `{{ ref('some_model') }}.is_view`.) 

If it doesn’t contain any special properties you need to have access to (and you can't infer the correct value in a develop context from their absence), then you could downcast the metric node to the same sort of object as you're creating from your develop macro, and pass those both into the get_metric_sql macro. This would enable that macro to not know/care how its inputs were created. 

In short, this:
```python
#macros/calculate.sql
 {%- set sql = metrics.get_metric_sql(
        metric_list=metric_list,
...
```
could instead be 
```python
#macros/calculate.sql
{% set metric_objects = metrics.get_metric_objects_from_names(metric_list) %}
{% set downcast_metric_objects = metrics.downcast_objects(metric_objects) %}

{% set sql = metrics.get_metric_sql(
        metric_list=downcast_metric_objects,
...
```
and 
```python 
#macros/develop.sql
    {% set develop_yml = fromyaml(develop_yml) %}
  # [lots of validation...]
     {% set downcast_metrics_objects = metrics.create_develop_downcast_objects(develop_yml) %}
{%- set sql = metrics.get_metric_sql(
        metric_list=downcast_metrics_objects,
...
```

Reply:
Hmmm. This **could** work but to me this feels more brittle. For multiple metrics, we'd need to create a list of dictionaries where each metric contains all of the components of the metric as additional key value pairs. 

IMO this might be better suited for a different issue focused on cleaning up `get_metric_sql` - we could pair it with the work of removing the filter for multiple metrics vs single metric. Directionally I'm aligned but I think that is internal optimization